lib/CFRPageActions.jsm

```diff
@@ -0,0 +1,225 @@
+const RecommendationMap = new WeakMap();
+
+/**
+ * A WeakMap from windows to their CFR PageAction.
+ */
+const PageActionMap = new WeakMap();
+
+/**
+ * We need one PageAction for each window
+ */
+class PageAction {
+  constructor(win) {
+    this.window = win;
+    this.urlbar = win.document.getElementById("urlbar");
+    this.container = win.document.getElementById("contextual-feature-recommendation");
+    this.button = win.document.getElementById("cfr-button");
+    this.label = win.document.getElementById("cfr-label");
+
+    this._popupStateChange = this._popupStateChange.bind(this);
+    this._collapse = this._collapse.bind(this);
+    this._handleClick = this._handleClick.bind(this);
+
+    // Saved timeout IDs for scheduled state changes, so they can be cancelled
+    this.stateTransitionTimeoutIDs = [];
+
+    this.container.onclick = this._handleClick;
+  }
+
+  show(notificationText, shouldExpand = false) {
+    this.container.hidden = false;
+
+    this.label.value = notificationText;
+
+    // We can safely get the width of the label even though the recommendation is
+    // collapsed; the label itself remains full width (with its overflow hidden)
+    const [{width}] = this.label.getClientRects();
+    this.urlbar.style.setProperty("--cfr-label-width", `${width}px`);
+
+    if (shouldExpand) {
+      this._clearScheduledStateChanges();
+
+      // After one second, expand
+      this._expand(1000);
+
+      // Five seconds later, collapse again
+      this._collapse(6000);
+    }
+  }
+
+  hide() {
+    this.container.hidden = true;
+    this._collapse();
+  }
+
+  _expand(delay = 0) {
+    if (!delay) {
+      // Non-delayed state change overrides any scheduled state changes
+      this._clearScheduledStateChanges();
+      this.urlbar.setAttribute("cfr-recommendation-state", "expanded");
+    } else {
+      this.stateTransitionTimeoutIDs.push(this.window.setTimeout(() => {
+        this.urlbar.setAttribute("cfr-recommendation-state", "expanded");
+      }, delay));
+    }
+  }
+
+  _collapse(delay = 0) {
+    if (!delay) {
+      // Non-delayed state change overrides any scheduled state changes
+      this._clearScheduledStateChanges();
+      this.urlbar.setAttribute("cfr-recommendation-state", "collapsed");
+    } else {
+      this.stateTransitionTimeoutIDs.push(this.window.setTimeout(() => {
+        this.urlbar.setAttribute("cfr-recommendation-state", "collapsed");
+      }, delay));
+    }
+  }
+
+  _clearScheduledStateChanges() {
+    while (this.stateTransitionTimeoutIDs.length > 0) {
+      // clearTimeout is safe even with invalid/expired IDs
+      this.window.clearTimeout(this.stateTransitionTimeoutIDs.pop());
+    }
+  }
+
+  // This is called when the popup closes as a result of interaction _outside_
+  // the popup, e.g. by hitting <esc>
+  _popupStateChange(state) {
+    if (["dismissed", "removed"].includes(state)) {
+      this._collapse();
+    }
+  }
+
+  _dispatchToActivityStream(action) {
+    // TODO
```

Comment:
This is actually going to be `dispatchToASRouter` now that I think about it. You could actually pass this in as a function in the constructor

Reply:
I've renamed it but haven't passed it into the constructor, though I can do if you'd prefer?